File: scala/26.remove-duplicates-from-sorted-array.scala
/*
 * @lc app=leetcode id=26 lang=scala
 *
 * [26] Remove Duplicates from Sorted Array
 */

// @lc code=start
object Solution {
    def removeDuplicates(nums: Array[Int]): Int = {

        // nums.tail contains all elements except the first operation
        // foldLeft iterates with the tuple as an initial value
        nums.tail.foldLeft((nums.head, 1)) {
            case ((i, j), current) =>
                if (current != i) {
                    nums(j) = current
                    (current, j + 1)
                } else {
                    (i, j)
                }
        }._2
    }
}
// @lc code=end

File: cpp/26.remove-duplicates-from-sorted-array.cpp
/*
 * @lc app=leetcode id=26 lang=cpp
 *
 * [26] Remove Duplicates from Sorted Array
 */

// @lc code=start
class Solution {
public:
    int removeDuplicates(vector<int>& nums) {
        // keep track of position of the next unique element
        int j = 1;

        // iterate from 1, because 0 is already a unique element
        for (int i = 1; i < nums.size(); i++) {

            // if current is not equal to previous, then it is unique
            if (nums[i] != nums[i - 1]) {

                // copy the unique element to the next position
                nums[j++] = nums[i];
            }
        }

        // return length of array with unique elements
        return j;
    }
};
// @lc code=end

